### PR TITLE
Centralize corner items

### PR DESCRIPTION
**Pull request by vitorandrietta**
refactor: centralize corner items

one way to centralize corner items is to add dummy items as new corner items, and after you scroll to it you can forcefully scroll back to a valid item.

with these changes when a user scrolls to one of the corners he will be scrolled back to the item that is directly before the dummy item.

### CODE REVIEW

File: src/carousel.js [hunk 1]
```diff
@@ -31,21 +32,29 @@ class Carousel extends Component {
...
     this.xOffset = new Animated.Value(0);
     this.halfContainerWidth = containerWidth / 2;
     this.halfItemWidth = itemWidth / 2;
+    this.alwaysCentralizeSelected = alwaysCentralizeSelected;
+    this.data = this.alwaysCentralizeSelected ?
+       [this.Constants.dummyItem, ...data,  this.Constants.dummyItem]: data
+    this.lastIndex = this.data.length - 1;
   }
-
+
```
Comment (joaobertacchi) [inline src/carousel.js]: nit: unnecessary change

File: src/carousel.js [hunk 2]
```diff
@@ -31,21 +32,29 @@ class Carousel extends Component {
...
-      initialIndex
+      initialIndex,
     } = this.props;
     this.currentIndex = initialIndex;
     this.scrollXBegin = 0;
     this.xOffset = new Animated.Value(0);
     this.halfContainerWidth = containerWidth / 2;
     this.halfItemWidth = itemWidth / 2;
+    this.alwaysCentralizeSelected = alwaysCentralizeSelected;
+    this.data = this.alwaysCentralizeSelected ?
```
Comment (joaobertacchi) [inline src/carousel.js:56]: if `this.props.data` changes, `this.data` will not reflect that. It might be a problem but depends on carousel usage...

File: src/carousel.js [hunk 3]
```diff
@@ -31,21 +32,29 @@ class Carousel extends Component {
...
+
   initialize() {
     const {
+      alwaysCentralizeSelected,
       itemWidth,
-      separatorWidth,
       data,
       containerWidth,
-      initialIndex
+      initialIndex,
```
Comment (YasminBeatriz) [inline src/carousel.js:46]: is this a bug you're fixing by adding the comma?  👀

File: src/carousel.js [hunk 4]
```diff
@@ -72,6 +80,9 @@ class Carousel extends Component {
     if (index < 0 || index >= this.data.length) return;
     onScrollEnd(this.data[index], index);
     this.currentIndex = index;
+    if(this.onItemSelected && (!this.alwaysCentralizeSelected || (index > 0 && index<this.lastIndex))){
```
Comment (YasminBeatriz) [inline src/carousel.js:85]: nit: format the syntax `index < this.lastIndex`

File: src/carousel.js [hunk 5]
```diff
@@ -260,6 +261,8 @@ class Carousel extends Component {
         onScroll={this.handleOnScroll}
         onScrollEndDrag={this.handleOnScrollEndDrag}
         getItemLayout={this.getItemLayout}
+        onTouchStart={alwaysSnapCenter ? this.handleOnScrollBeginDrag: null}
```
Comment (YasminBeatriz) [inline src/carousel.js]: nit: format the syntax: `this.handleOnScrollBeginDrag : null`

File: src/carousel.js [hunk 6]
```diff
@@ -260,6 +261,8 @@ class Carousel extends Component {
...
+        onTouchEnd={alwaysSnapCenter ? this.handleOnScrollEndDrag: null}
```
Comment (YasminBeatriz) [inline src/carousel.js]: ditto

File: src/carousel.js (hunk 2 above)
Comment (joaobertacchi) [reply src/carousel.js:56]: I was about to suggest you to use `componentWillReceiveProps`. However it's now deprecated (https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops) and you should use a memoization as described at https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization.

File: src/carousel.js [hunk 7]
```diff
@@ -179,25 +202,28 @@ class Carousel extends Component {
...
-  getItemLayout(data, index) {
-    const {itemWidth,separatorWidth}=this.props;
-    return {
-      offset:index * (itemWidth + separatorWidth) +
-          this.halfItemWidth -
-          this.halfContainerWidth,
-      length: itemWidth,
-      index
-    };
+  getItemLayout(data, index) {
```
Comment (joaobertacchi) [inline src/carousel.js:233]: Avoid unnecessary styling changes...

File: src/carousel.js [hunk 8]
```diff
@@ -59,9 +68,9 @@ class Carousel extends Component {
   }
 
   scrollToIndex(index) {
```
Comment (joaobertacchi) [inline src/carousel.js:79]: it should not allow to scroll to `index=0` or `index=data.length-1` when `alwaysCentralizeSelected` is true.

File: src/carousel.js [hunk 9]
```diff
@@ -94,11 +103,25 @@ class Carousel extends Component {
     scrollDistance < 0
```
Comment (joaobertacchi) [inline src/carousel.js]: Let's say that `scrollDistance < 0`, `this.currentIndex===1` and `this.alwaysCentralizeSelected===true`. `this.scrollToIndex(0)` will be called and will change `this.currentIndex` to `0`. Then `this.scrollToIndex(1)` will be called.
`scrollToIndex` should be called only once when drag ends...

File: src/carousel.js (hunk 2 above)
Comment (vitorandrietta) [reply src/carousel.js:56]: I'm prioritizing fixing the scroll behavior, I will fix it after.

File: src/carousel.js (hunk 9 above)
Comment (vitorandrietta) [reply src/carousel.js]: this was an intended behavior in order to display an automatic (drag back effect).
but i agree, it's not the best possible impl.

now it's fixed by adding the following validation to `scrollToIndex` 
`if(this.alwaysCentralizeSelected  && (index === 0 || index === this.data.length-1)) 
return;`
